### PR TITLE
Name and symbol now determined at creation-time using SequenceRegistr…

### PR DESCRIPTION
**Pull request by hoytech**
…y, and cannot be changed afterwards

### CODE REVIEW

File: src/SequenceRegistry/SequenceRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+/// @notice This contract maintains sequence counters for any/all addresses, each starting at 1
+/// @dev Anybody can reserve a sequence ID. The only guarantee provided is that no two reservations for the same asset will get the same ID.
+contract SequenceRegistry is ISequenceRegistry {
+    /// @dev Each asset maps to the last sequence ID issued, or 0 if none were ever issued.
+    mapping(address asset => uint256 lastSeqId) public counters;
+
+    /// @notice Reserve an ID for a given asset
+    /// @param asset The address that this ID is for
+    /// @return seqId Sequence ID
+    function reserveSeqId(address asset) external returns (uint256 seqId) {
```
Comment (kasperpawlowski) [inline src/SequenceRegistry/SequenceRegistry.sol:18]: nit: usually, we don't do named returns unless it improves clarity. there was a lot of discussions about it with Spearbit

File: src/SequenceRegistry/SequenceRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import {ISequenceRegistry} from "../interfaces/ISequenceRegistry.sol";
+
+/// @title SequenceRegistry
+/// @author Euler Labs (https://www.eulerlabs.com/)
+/// @notice This contract maintains sequence counters for any/all addresses, each starting at 1
+/// @dev Anybody can reserve a sequence ID. The only guarantee provided is that no two reservations for the same asset will get the same ID.
+contract SequenceRegistry is ISequenceRegistry {
+    /// @dev Each asset maps to the last sequence ID issued, or 0 if none were ever issued.
+    mapping(address asset => uint256 lastSeqId) public counters;
```
[the remark this answers is not in the document]
Comment (hoytech) [reply src/SequenceRegistry/SequenceRegistry.sol]: I don't really think anyone will ever use this. No need for it to be part of the interface.

File: src/EVault/shared/Base.sol [hunk 3]
```diff
@@ -19,18 +20,21 @@ import "./types/Types.sol";
...
         address evc;
         address protocolConfig;
+        address sequenceRegistry;
         address balanceTracker;
         address permit2;
     }
 
     constructor(Integrations memory integrations) EVCClient(integrations.evc) {
         protocolConfig = IProtocolConfig(AddressUtils.checkContract(integrations.protocolConfig));
+        sequenceRegistry = ISequenceRegistry(integrations.sequenceRegistry);
```
Comment (kasperpawlowski) [inline src/EVault/shared/Base.sol:37]: use `AddressUtils.checkContract` to verify the sequenceRegistry is a contract

File: src/EVault/modules/Initialize.sol [hunk 4]
```diff
@@ -47,6 +47,14 @@ abstract contract InitializeModule is IInitialize, Base, BorrowUtils {
         vaultStorage.interestFee = DEFAULT_INTEREST_FEE.toConfigAmount();
```
Comment (kasperpawlowski) [inline src/EVault/modules/Initialize.sol:47]: let's use `AddressUtils.checkContract` for the line above too:

`if (address(asset).code.length == 0) revert E_BadAddress();`

**Review by kasperpawlowski — APPROVED** [review]

File: src/EVault/modules/Initialize.sol [hunk 5]
```diff
@@ -47,6 +47,14 @@ abstract contract InitializeModule is IInitialize, Base, BorrowUtils {
...
         vaultStorage.creator = vaultStorage.governorAdmin = proxyCreator;
 
+        {
+            string memory underlyingSymbol = getTokenSymbol(address(asset));
+            uint256 seqId = sequenceRegistry.reserveSeqId(underlyingSymbol);
+
+            vaultStorage.symbol = string(abi.encodePacked("e", underlyingSymbol, "-", uintToString(seqId)));
```
Comment (mario-eth) [inline src/EVault/modules/Initialize.sol:54]: I think https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Strings.sol#L24 OZ version might be more gas optimized?

**Pull request merged**